src/get-started/cryptocurrency-advanced.md

```diff
@@ -0,0 +1,835 @@
+      wallet: Wallet,
+      amount: u64,
+      transaction: &Hash
+    ) {
+        let wallet = {
+            let mut history = self.wallet_history_mut(wallet.pub_key());
+            history.push(*transaction);
+            let history_hash = history.merkle_root();
+            let balance = wallet.balance();
+            wallet.set_balance(balance - amount, &history_hash)
+        };
+        self.wallets_mut().put(wallet.pub_key(), wallet.clone());
+    }
+
+    /// Creates new wallet and appends first record to its history.
+    pub fn create_wallet(
+      &mut self,
+      key: &PublicKey,
+      name: &str,
+      transaction: &Hash
+    ) {
+        let wallet = {
+            let mut history = self.wallet_history_mut(key);
+            history.push(*transaction);
+            let history_hash = history.merkle_root();
+            Wallet::new(key, name, INITIAL_BALANCE, history.len(), &history_hash)
+        };
+        self.wallets_mut().put(key, wallet);
+    }
+}
+```
+
+## Reporting Errors
+
+Before defining the types of transactions, the service will realize, we define
+the types of errors that might occur
+during execution thereof. The section is identical to the
+[one]( https://exonum.com/doc/get-started/create-service/#reporting-errors) in
+the simple Cryptocurrency demo.
+
+You might have noticed that for this purpose we use an external crate `failure`
+that was mentioned in the dependencies
+above. The errors list covers typical cases a user might encounter during
+service application that are quite self-explanatory:
+
+```rust
+/// Error codes emitted by wallet transactions during execution.
+#[derive(Debug, Fail)]
+#[repr(u8)]
+pub enum Error {
+    /// Wallet already exists.
+    ///
+    /// Can be emitted by `CreateWallet`.
+    #[fail(display = "Wallet already exists")]
+    WalletAlreadyExists = 0,
+
+    /// Sender doesn't exist.
+    ///
+    /// Can be emitted by `Transfer`.
+    #[fail(display = "Sender doesn't exist")]
+    SenderNotFound = 1,
+
+    /// Receiver doesn't exist.
+    ///
+    /// Can be emitted by `Transfer` or `Issue`.
+    #[fail(display = "Receiver doesn't exist")]
+    ReceiverNotFound = 2,
+
+    /// Insufficient currency amount.
+    ///
+    /// Can be emitted by `Transfer`.
+    #[fail(display = "Insufficient currency amount")]
+    InsufficientCurrencyAmount = 3,
+}
+
+// Conversion between service-specific errors and the standard error type
+// that can be emitted by transactions.
+impl From<Error> for ExecutionError {
+    fn from(value: Error) -> ExecutionError {
+        let description = format!("{}", value);
+        ExecutionError::with_description(value as u8, description)
+    }
+}
+```
+
+The implementation code above returns a JSON that provides a corresponding error
+code together with its verbal description.
+
+## Define Transactions
+
+We use transactions! macro to define the service transactions.
+
+In this tutorial, however, we unite the transactions under the
```

Comment:
I don't quite understand the connotation of "however" here. What is this code contrasted with?